src/lsp-extensions.ts

```diff
@@ -0,0 +1,39 @@
+import * as lsp from 'vscode-languageserver/node';
+import { RequestHandler } from 'vscode-jsonrpc';
+
+export namespace extensions.inlayHints {
+    export type InlayHintsParams = {
+        /**
+         * The document to format.
+         */
+        textDocument: lsp.TextDocumentIdentifier;
+        /**
+         * The range to format
+         */
+        range?: lsp.Range;
+    };
+
+    type InlayHint = {
+        text: string;
+        position: lsp.Position;
+        kind?: 'Type' | 'Parameter' | 'Enum';
+        whitespaceBefore?: boolean;
+        whitespaceAfter?: boolean;
+    };
+
+    export type InlayHintsResult = {
+        inlayHints: InlayHint[];
+    };
+
+    export const type = new lsp.RequestType<
+    InlayHintsParams,
+    InlayHintsResult,
+    lsp.TextDocumentRegistrationOptions
+    >('typescript/inlayHints');
```

Comment:
Have you implemented this based on existing implementation or some specific spec? I know there is attempt to put that feature into the official spec (https://github.com/microsoft/language-server-protocol/pull/1249/files) but the one you have here is not that.

I'm fine with implementing this in the meantime since it uses a custom request. That's not a problem. The only thing I wonder is why is the result an object with `inlayHints` property on it. The result should likely itself just be an array of `InlineHint`.

Reply:
I just used something that is more or less equal to what typescript provides as a custom request. We can later migrate to the official spec when it's fully defined.

As for the return type, in general, it's bad practice to return arrays because it will be impossible to return more than the inlayHints if we would ever want that. Now we can just add an extra field.

I can always change if you think it's better to return an array.

Reply:
> As for the return type, in general, it's bad practice to return arrays because it will be impossible to return more than the inlayHints if we would ever want that. Now we can just add an extra field.

OK. It's an experimental API anyway so it doesn't really matter long-term.